### PR TITLE
Start of new examples app for integration tests

### PR DESCRIPTION
**Pull request by GGAlanSmithee**
References: https://github.com/styled-components/styled-components/issues/1358 and https://spectrum.chat/?t=d475965b-2031-40c0-90e5-bb8159d46877

This pretty much removes the old examples app and replaces it with a new one, which is more adapted towards integration tests. The new one, in its current state, is a very simple framework for adding examples (for each `styled-components` API / export). The only API example currently included is the `css` one.

Notable changes:
* refactored examples app
    * removed all ssr and stresstests
    * introduced a simple framework for adding examples
    * removed all third party scripts in favor of a bundle (see below)
* added bundling for the examples app
    * using `rollup` as with all other builds
    * this was a reasonable tradeoff to me, since getting jest to play nice with SSR was a hassle
* removed the `dev` script in favor of `build:example` and `run:example` scripts
* updated integration tests

Remarks / Thoughts
* `build:example` takes ~10 sec and there is no dev server.
    * Can be improved by excluding 3rd party libs
    * Can add dev server if needed
* Moving forward, we can do one of two things, as I see it:
    1. include all examples in one page (can jest handle this?)
    2. multiple pages, with one example per page and with one corresponding test per page
    3. since it's only used for integration tests now, should it be moved there, or kept as /example?

cc @mxstbr @dreamorosi @briandridge and also @philpl regarding test structuring

Let me know if this is going in the right direction, or if you are looking for something else completely.

### CODE REVIEW

**Comment by philpl** [comment]
Thanks for picking this up! :heart: So unfortunately there's a lot of problems here. I previously wanted to pick up the `example/` setup so that I can improve it before we try to implement new example components. Optimally this should be split into two PRs.

I'll just try to list out the problems I see and hopefully we can converge onto what we're searching for. Don't worry if not, I can always pick the setup part of it up once I find the time... So here goes nothing:

- The example bundle should not be committed to git
- The example bundle should not be built using a build step / script / as part of the release process
- We do need SSR since that's an important part to test
- Webpack should be used; we can just use webpack-hot- and webpack-dev-middleware
- We might not need bundle/prod builds
- Optimally, SSR+the app should not use two webpack instances
- Optimally, SSR and the rest of the webpack server/build should not use two processes
- All the dependencies and scripts to run the example should only live in `example/` i.e. nothing in the rest of the repo should be polluted by it

I know this is quite a bit to drop onto you, but I needed to write this down either way :/

Let me know what you think, how I can help, and how you think we should proceed :smile:

**Comment by GGAlanSmithee** [comment]
Hi @philpl thanks for the feedback, no problem that it's a lot to change, let's fix it

* Originally, I went for ssr because of your reasoning, this can probably be fixed, but I hit github.com/styled-components/styled-components/issues/893 which I tried to solve using `StyleSheet.reset`, but could not get it to work, should I use https://github.com/styled-components/jest-styled-components ?
* Regarding building / bundling etc, that's no problem, I'll change it. I jacked in to your current build setup to avoid introducing more deps
* Do we need client side hydration, or is it enough to render a static page using ssr?

So I'll start fixing this, let's see how it goes :)

**Comment by GGAlanSmithee** [comment]
@philpl I have reverted a lot of the changes;

* now there is only ssr (no client side rendering at this moment)
* there is no bundling (since we only render server-side using babel-noe atm
* I experimented with ssr so now the integration-tests are working
    * Not sure if I'm doing it the correct way

The PR should be much easier to grasp and evaluate now, please let me know if it's going in the right direction.

File: rollup.config.js [hunk 1]
```diff
@@ -47,18 +47,19 @@ const plugins = [
   commonjs({
```
Comment (GGAlanSmithee) [inline rollup.config.js]: this is auto-formatted, should be no changes from me here

**Comment by GGAlanSmithee** [comment]
So, the reason this PR is failing CI tests is because of the precommit hook which autoformats `rollup-config.js`, introducing a dangling comma on [line 53](https://github.com/styled-components/styled-components/pull/1375/files#diff-ff6e5f22a9c7e66987b19c0199636480R53). Not sure how to fix this though, cc @mxstbr

**Comment by philpl** [comment]
@GGAlanSmithee We're fixing the hook in a separate PR; 

Regarding the new setup, we'd optimally need SSR+Client Rehydration too.
The rollup config should also not be changed, since it's unrelated to the example, is going to change in another PR, and the server should be just node.js code that uses the client code that can be compiled to a bundle with webpack, that also emits the prepared code for the server

**Comment by GGAlanSmithee** [comment]
@philpl thanks for your feedback;

> The rollup config should also not be changed

The rollup config isn't changed by me, so that should be resolved once the hook is fixed.

>  server should be just node.js code that uses the client code that can be compiled to a bundle with webpack,

Okay, I've set up a few projects using this stack, so should be no problem, Will take a look at your project too. Will add `project.json`, `webpack.config.js` etc to `/example` then, to keep it out of the root, as requested

**Comment by philpl** [comment]
@GGAlanSmithee hm, not sure why it's reformatting all of a sudden; the lint-staged target is probably off

Btw it looks like the entire `yarn.lock` file was recreated?

**Comment by gribnoysup** [comment]
I know that the decision at the moment is to use `example` app. But maybe the `sandbox` can be used for this? It has the setup ready for this already with ssr and rehydration. We can introduce some kind of routing to the sandbox, add a menu with different examples, etc. Then you can edit them while testing changes locally and use the same app for the screenshot testing

**Comment by GGAlanSmithee** [comment]
@philpl I will reset `yarn.lock`

@gribnoysup yeah, looks like the sandbox app is closer to what we are looking for. Might make more sense to use that for this purpose too. Then we should probably have one example per route, which would make it easier to see which tests is failing, etc.

**Comment by mxstbr** [comment]
That makes a lot of sense to me, I don't see why we _need_ to have two apps in the repo tbch

**Comment by gribnoysup** [comment]
@ggalansmithee yes, I was thinking of something like this! :)

**Comment by GGAlanSmithee** [comment]
I will wait for some more input (@philpl) then I will close this PR and start working on the sandboax instead. :+1:

**Comment by philpl** [comment]
@GGAlanSmithee That sounds like the way to go; A lot of time has been poured into that already, so it'd make sense to just combine the two :+1:

**Comment by GGAlanSmithee** [comment]
Closing this as per the above conversation, will move forward with the sandbox app instead

**Comment by gribnoysup** [comment]
Just a note, I'm working on the slight refactor for the sandbox server that @philpl suggested in the https://github.com/styled-components/styled-components/pull/1257#issuecomment-351054607. I'm struggling a little with HMR at the moment but overall it works much better and without these `eval` hacks. I think I'm close to finishing it soon and will create a PR for this

@GGAlanSmithee, the `src` folder that contains all the client code in the sandbox wouldn't change at all, so you can start with adding examples right away, the base setup is ready to be used for the screenshot testing, I think. And with this approach, we probably won't even have any conflicts when I'm finished! Feel free to ping me at any time if you need any help with the sandbox setup as I'm the one responsible for this code 😅

**Comment by GGAlanSmithee** [comment]
@gribnoysup sounds like a plan :+1: I'll have a look and start expermineting with a simple example, some form of routing and pointing the integration tests towards this app.

**Comment by GGAlanSmithee** [comment]
@gribnoysup @philpl thinking about react-router for routing, mostly because I'm very used to it. This will require us to change the eval of the `server.js` script though, to have it work on the server. Any thoughts on this? Will this conflict with what you are doing?

**Comment by mxstbr** [comment]
Why are we doing multiple pages again? Can't we just put everything on one page:

**Comment by GGAlanSmithee** [comment]
@mxstbr 

I guess we could. There are some things to take into consideration though.
* It will be on the same page as the sandbox
* Screenshot test will have to handle scrolling
* All examples will be in the same test (unless it's configurable) which will make it harder to determine which one failed

Also, the `eval` solution currently implemented for the server is a bit.. weird.

**Comment by mxstbr** [comment]
> All examples will be in the same test (unless it's configurable) which will make it harder to determine which one failed

Yeah that's a good point, it _would_ be nicer to get a clear message which specific stuff broke 🤔 

I guess just start out without `react-router` server-side, and then implement it once @gribnoysup has done his updates?

**Comment by GGAlanSmithee** [comment]
@mxstbr sounds like a plan 👍

**Comment by gribnoysup** [comment]
What is the problem of using RR right away? You can change the `server.js` or any other code in the sandbox, I'll make appropriate adjustments in my code, don't worry about the conflicts 😉

**Comment by GGAlanSmithee** [comment]
@gribnoysup no problem other than the eval part of the server-side rendering. I'll fix it.

**Comment by gribnoysup** [comment]
@GGAlanSmithee I'm almost done with changing the bootstrap code for the sandbox, but if you still want to proceed on your own, you can refer to the @philpl comment I mentioned above and I will stop working on this.

**Comment by GGAlanSmithee** [comment]
@gribnoysup no, by all means continue with your changes, I'll just merge as I go.

**Comment by albinotonnina** [comment]
Hello everyone! Is someone currently working on this one?

**Pull request closed without merging**